Userland/Libraries/LibJS/SourceTextModule.cpp

```diff
@@ -19,18 +81,571 @@ Result<NonnullRefPtr<SourceTextModule>, Vector<Parser::Error>> SourceTextModule:
+ThrowCompletionOr<Value> SourceTextModule::execute_module(VM& vm, Optional<PromiseCapability> capability)
+{
+    dbgln_if(JS_MODULE_DEBUG, "[JS MODULE] SourceTextModule::execute_module({}, capability has value: {})", filename(), capability.has_value());
+
+    // 1. Let moduleContext be a new ECMAScript code execution context.
+    ExecutionContext module_context { vm.heap() };
+
+    // Note: This is not in the spec but we require it.
+    module_context.is_strict_mode = true;
+
+    // 2. Set the Function of moduleContext to null.
+
+    // 3. Set the Realm of moduleContext to module.[[Realm]].
+    module_context.realm = &realm();
+
+    // 4. Set the ScriptOrModule of moduleContext to module.
+    module_context.script_or_module = this;
+
+    // 5. Assert: module has been linked and declarations in its module environment have been instantiated.
+    VERIFY(m_status != ModuleStatus::Unlinked && m_status != ModuleStatus::Linking && environment());
+
+    // 6. Set the VariableEnvironment of moduleContext to module.[[Environment]].
+    module_context.variable_environment = environment();
+
+    // 7. Set the LexicalEnvironment of moduleContext to module.[[Environment]].
+    module_context.lexical_environment = environment();
+
+    // 8. Suspend the currently running execution context.
+    // FIXME: We don't have suspend yet
+
+    // 9. If module.[[HasTLA]] is false, then
+    if (!m_has_TLA) {
+        // a. Assert: capability is not present.
+        VERIFY(!capability.has_value());
+        // b. Push moduleContext onto the execution context stack; moduleContext is now the running execution context.
+        vm.push_execution_context(module_context, realm().global_object());
+
+        // c. Let result be the result of evaluating module.[[ECMAScriptCode]].
+        auto result = m_ecmascript_code->execute(vm.interpreter(), realm().global_object());
+
+        // d. Suspend moduleContext and remove it from the execution context stack.
+        vm.pop_execution_context();
+        vm.clear_exception();
+
+        // e. Resume the context that is now on the top of the execution context stack as the running execution context.
+        // FIXME: We don't have resume yet.
+
+        // f. Return Completion(result).
+        if (result.is_error())
+            return result;
+
+        // 16.2.1.11 Runtime Semantics: Evaluation, https://tc39.es/ecma262/#sec-module-semantics-runtime-semantics-evaluation
+        // -> Replace any empty value with undefined.
+
+        result = result.update_empty(js_undefined());
+        return *result.value();
+    }
+    // 10. Else,
+
+    // a. Assert: capability is a PromiseCapability Record.
+    VERIFY(capability.has_value());
+
+    // b. Perform ! AsyncBlockStart(capability, module.[[ECMAScriptCode]], moduleContext).
+    async_block_start(vm, m_ecmascript_code, capability.value(), module_context);
+
+    // c. Return NormalCompletion(empty).
+    return js_undefined();
```

Comment:
I'm afraid this function will need to return a proper Completion to represent this return state - undefined is not the same as empty.